### ISSUE TITLE
Dashboard freeze "infinitely" when template variable selected, when time range change or dashboard refreshed

### ISSUE THREAD

**Issue by jtorkkel** [grafana/grafana#46510]
**What happened**:
When dashboard is started all template variables are loaded fast in ~1s (first picture), sometimes might take longer due few larger queries.

Dashboard freeze for few minutes when time range is changed (5m --> 15m) or dashboard is reloaded. 
Dashboard freeze for 1min in one variable list is opened and one variable is selected (variable used on most panels)
Template variables animation "in progress" is shown until query completed. Sometimes few variable stay in "in progress" for 10-15 minutes.

Dashboards contains ~40 templated chained variable. All dependent on job (loaded on dashboard load), some also dependent on other few variables which are used for further filtering. All other variables are refreshed on time range change.

Dashboard panels are located under tens of rows, each rows contain one or more repeating panel utilizing templated variables.
All panels are located inside collapsing rows which often use repeating panels.

All panels are under rows to minimize unnecessary loads while updating time range. Naturally only few are expanded on need based.

Grafana version 8.2.15 works, no freeze. 
Grafana version 8.3.4 - grafana-8.5.0-54667pre, freeze
Grafana 8.3.0Beta1-8.3.3 cannot be properly tested as dashboards load fails to error. Simpler dashboards works
```
Templating init failed Minified Redux error #3; visit https://redux.js.org/Errors?code=3 for the full message or use the non-minified dev environment for full errors.
```

Using inspector/Network the following pattern can be seen.
- complete freeze 10-30s, burst of 20-30 template variable queries in pending (10s-30s)
- After 10-30s the queries exit pending state and completed in ~100ms
- 1-4 new template variable queries in pending state for 10-30s
- After 10-30s the queries exit pending state and completed in ~100ms
- next 1-4 queries template variable 
- until all template variables queried
Some queries fails, when takes too long time. Failure is not due long response time from Prometheus but probably freeze interrupt the query again.

When variable is selected the above pattern repeats 2 times (thus UI freeze for 1min), because only few variables get refreshed I think.
When time range change or dashboard is refreshed the freeze takes 5min because all variables are reloaded.
 
Dashboard launched (~1s), always fast-
![image](https://user-images.githubusercontent.com/19704268/158062205-1928355c-1c9a-4bd9-bcdd-af6b58d75137.png)

One variable selected (instead of default all, freeze for 1min. Picture shows first everything freeze for 20s)
![image](https://user-images.githubusercontent.com/19704268/158062298-193be5f6-cda0-4ab7-ade5-3b1c17e26be8.png)

Dashboard refresh few second after launch
![image](https://user-images.githubusercontent.com/19704268/158063114-3ece4cc6-2a6a-47a0-8dcf-0f83454e21e9.png)

After 3min
![image](https://user-images.githubusercontent.com/19704268/158063219-fe166d43-edd9-45c1-8bb4-f087333cf38a.png)

After 5min queries completed
![image](https://user-images.githubusercontent.com/19704268/158063310-1ca318f4-6a1e-4376-b949-9ce9a7933c1d.png)

Inspector/Performance indicate all +99% time spent on scripting.
Inspector cannot be started during freeze, must be started before, takes very long time to stop after 5min.
Result looks like this 2min after pressing refresh (Windows 8.4.2, separate take, takes 10m to stop, after 5min will never stop)
![image](https://user-images.githubusercontent.com/19704268/158069595-09ea34cc-276d-4ba8-8ee8-0e40bc0225b4.png)

![image](https://user-images.githubusercontent.com/19704268/158069742-44e10a8a-37c7-4774-9045-150d6b4ad498.png)

**What you expected to happen**:
- All variables should be loaded before starting to refresh panels, should not take more than few second
- Panels under collapsed rows must not be refreshed.
- if some panels are not under collapsing row or row is open the panels should be refreshed only after variables are updated.

**How to reproduce it (as minimally and precisely as possible)**:
Can be simulated with datasource which contains enough metrics/cardinality/labels. Just create tens of "dependent" template variables and tens of rows with collapsing dashboard each using different template variable sets.
Issue can be seen also on smaller dashboards but seems that the larger the number of template/queries under rows the slower and longer the freeze.
I did not find a way to query variables from testData DB.

I suspect this is regression due
Variables: Only update panels that are impacted by variable change. https://github.com/grafana/grafana/pull/39420, [@hugohaggmark](https://github.com/hugohaggmark)

What this PR does / why we need it:
This PR tries to resolve what panels are impacted during a variable change. It takes a naive approach to chained variables and parses the dashboard json in order to resolve the dependencies. This means if a variable query is faulty or if the user cancels the variable query the affected panels for that variable will still be refreshed.

It looks that 
- dashboard json is analyzed too early breaking variable loading
- analysis is too compute intensive eating all threads
- analysis is done done multiple times 
- does not make any sense to analyze dependencies when rows are collapsed.

Dashboards with tens of template variables and tens of rows with collapsing panels could have hundreds or thousands of queries with template variables to be analyzed, might require scanning JSON once per template variable to see if refresh is needed.

I tested another simpler production dashboards which aggregates the alerts from selected time range.
Dashboard has 25 "dependent" template variables which are used to list only variables which are part of alerts and 15 rows with collapsing panels showing then all alerting dashboards. As seen from graph after few variables queried the is 1s freeze between. 1s freeze is probably due json scanning. This hints that quite simple dashboard from data source with large amount of metrics regress clearly. And larger dashboards slow down exponentially making dashboards unusable.

![image](https://user-images.githubusercontent.com/19704268/158063878-280c8ae6-ef35-4168-8797-1bd7cc180e1f.png)

Same dashboard refresh when using 8.2.7 (one query is longer and some other are dependent on it)
![image](https://user-images.githubusercontent.com/19704268/158064280-09ca5456-acea-42a3-beb0-35fe92dffa34.png)

We had to revert 8.3/8.4 Grafana back to 7.5.15 due freeze. Probably due breaking change "Variables: Do not save repeated panels, rows and scopedVars. https://github.com/grafana/grafana/pull/32436, [@torkelo](https://github.com/torkelo)" some dashboards lost their Prometheus data source when opened in 7.5. Will try to fix by migrating to 8.2.15.

**Anything else we need to know?**:

**Environment**:
- Grafana version: 8.3.x/8.4.x/8.5.x
- Data source type & version: Prometheus
- OS Grafana is installed on: RHEL 7.9, WIndows 10
- User OS & Browser: Chrome
- Grafana plugins: Prometheus
- Others:

**Comment by joshhunt** [comment]
Thanks for reporting this with such detail! We're investigating this, and will keep the issue updated with notes along the way.

To clarify - the freezes you notice only happen when changing a template variable or clicking the "Refresh Dashboard" button? When you first load the dashboard (click your browser's refresh button) it loads fine?

Would it be possible to share a screenshot of the variable dependency graph (Dashboard Settings > Variables > Show dependencies)? Or, could you describe the length of your variable chains - how many variables are depending on each other?

---

Notes:
 - Attempting to reproduce the issue in [this Play dashboard](https://play.grafana.org/d/aDzRNDP7z/46510-lots-of-variable-freeze), but have not get got ~40 variables in.
 - It appears panels that are under repeated rows are always updated when any variable is changed (including one which is not in the panel, or depended on by the variables in the panel)

**Comment by torkelo** [comment]
40 variables?? can you include the list of variables and their dependencies? so we can try to build a similar dashboard to try to replicate this

**Comment by jtorkkel** [comment]
I got 8.3.2 working by touching "job" variable from settings.
Now I can see that 8.3.2 also freeze but for shorter time (3s), total 30s than 8.3.4. I could not find nightly 8.0 builds so cannot verify exact which nightly 8.0.0 build cause first regression.

Probably there is two regression first https://github.com/grafana/grafana/pull/39420 in 8.0.0beta and then https://github.com/grafana/grafana/pull/42828 or https://github.com/grafana/grafana/issues/43843 in 8.3.3

Panels under rows should not be touched or should be touched very carefully.

**Comment by axelavargas** [comment]
Thanks for the additional info @jtorkkel 🙌🏾, we were able to reproduce the issue during our internal mob session, we'll investigate further.

---

Notes:

- Issue reproduced on [this Play panel](https://play.grafana.org/d/aDzRNDP7z/46510-lots-of-variable-freeze), with ~40 variables on.
- After a quick look at the profiling file, we noticed that `variablesChangedHandler` is being fired hundreds of times when we chain variables and use them in repeats (rows and panels), that function is triggering the DashboardModel `startRefresh` function and that it eventually triggers an update on each panel

**Comment by ashharrison90** [comment]
ok had to go back to v8.2.7 to compare but i think i've found _something_.

as far as i can tell, in our example dashboard here: https://play.grafana.org/d/aDzRNDP7z/46510-lots-of-variable-freeze the vast majority of time is spent mounting/unmounting react panels. templating actually plays a very small part. 

i think we've introduced a performance regression with the introduction of `cleanUpPanelState` in https://github.com/grafana/grafana/pull/40302 cc @torkelo 

in `v8.2.7` there is no `cleanUpPanelState`, so the react internal function `commitUnmount` takes anywhere from 0.1-0.5ms per panel ⚡ 
![image](https://user-images.githubusercontent.com/20999846/160853300-8a5bbbca-8fda-45e4-83c0-f71c1668aaa9.png)

in `v8.5.0-pre (8426cfe400)` (and i suspect every version after https://github.com/grafana/grafana/pull/40302) it's taking between 8-15ms 🐌  mostly due to the new `cleanUpPanelState` calls when unmounting the panels
![image](https://user-images.githubusercontent.com/20999846/160854479-c57e5e4f-3b20-4847-92a5-0d9739bb552a.png)

with our example dashboard, we've got 55*55 = 3025 panels.
- clean up before: 3025 * ~0.3 = ~0.9075s
- clean up after: 3025 * ~11 = ~33.275s

now tbh a dashboard with 3000 panels is pretty unusable so this is an extreme case.

@jtorkkel some further questions because i'm not sure this is accurately replicating the issue you're seeing:
- can you confirm how many panels are being rendered on the dashboard?
- it would be good to know if your dashboard is also spending a majority of the time rendering panels. here you can see ours spent 50s in the `processRepeats` function, almost all of which was spent doing react internal things (`Component.forceUpdate`).
![image](https://user-images.githubusercontent.com/20999846/160858378-1510f2d0-fbd5-46e9-a291-919f46f35f32.png)
- you mentioned collapsing rows didn't help, but in our case this substantially improved the performance
- what hardware are you using to browse grafana? i notice your memory heap only ever goes up to around 250mb, whereas with the example dashboard above mine's hit >1gb on a m1 macbook 😅

**Comment by jtorkkel** [comment]
Hi, 
timeframe of 40302 could match first 8.3.0 release.

I have all collapsing rows closed, so it should not render any panels. 
Apparently cleanup consume all resources/threads and slow down variable reload.
Once variables are reloaded dashboards works normally. 
If variable value is changed when rows collapsed it also cause shorter freeze (variables are not reloaded)

I cannot see any symbols, so cannot see what is happening. All I can see is screenshot I put into error. 
Would nightly build contain symbols ?

My dashboards has more rows/panels than yours, thus it freeze for 5min I think.
With smaller dashboards freeze for shorter time but any decent complexity size dashboard is unusable.

I have Lenovo P53 i7 9850H 6 core/12 threads 32GB

**Comment by ashharrison90** [comment]
@jtorkkel your dashboards have more than 3000 panels? 😅 

is breaking them down into smaller dashboards an option? i would say 3000 panels on an individual dashboard is much more than decently complex. 😄

**Comment by jtorkkel** [comment]
The large one is master dashboards which I break to smaller ones already, but the smaller one are also freezing for 10s-2min.
Not possible to have such small dashboards on large system which would not freeze.
Small dashboards possible just for visualizing on system level, but for alerting and deep diving you need to identify first the problem and then deep dive to few selected ready made dashboards. Bu you need to have all possible dashboards available.

The large number of template variable multiplied with large number of hidden panels creates the issue.

Just imagine hundreds of uservices running on 100 hosts, few 1 to N instance each. uservices created own mesh and also connect to hundreds of other external services. You collect uservice metrics, metrics between uservices, metrics to external systems, host metrics, container metrics. Millions on timeseries.

You create tens of different panels containing all kind of panels for single uservice. There are many of those. You put them under different rows, each row when expanded render 1-5 panels for 1 to N uservices. uservices to be rendered selected through variable. Disabling variable option "All" prevent accidentally opening row which would render all 300 uservices.

System level panels (not repeating but summing) which contains similar variable with "all" when you want to aggregate metrics for all services. Same for host metrics etc.
Option to limit the number of repeating panels to 4-8 on some panels would reduce the number of theoretical panels.

Visualizing all kind of alerts also require quite many panels with all options, when no alerts nothing to show but again theoretical number is large.

So if all rows expanded with max number of variables then panel number would be large. But never all expanded just few depending on need base.

Few system panels contains still "all" option and maybe many repeating panels (should not be opened when all selected) which might explode the number of theoretical panels. I can create yet another variable for those too and disable option all. That would decrease the number of theoretical panels by few thousands and might result 2x shorter freeze, but would not fix the freeze. Would help on shortening the freeze when exiting dashboard.

**Comment by aengusrooneygrafana** [comment]
Unfortunately https://github.com/grafana/grafana/pull/47323 did not improve the freeze in 8.5.0

**Comment by jtorkkel** [comment]
In playground rendering is slow, but I suspect that template variables are not refreshed on time range change as not progress circle seen on variables when 1y range selected. No difference if range is 5min vs 1 year.

**Comment by natellium** [comment]
@jtorkkel We introduced some more improvements that will go out in 9.0.5.https://github.com/grafana/grafana/pull/52047
 Could you let us know if this helped?

**Comment by jtorkkel** [comment]
I made quick tests using 9.0.5, now dashboard is not freezing. Template variable loading time is maybe 20-50% longer than in 8.2.7 for short ranges like 30min. Now dashboard is usable.

Dashboard with 50 template variables result roughly 120 queries in 8.2.7 and 9.0.5. Optimally only 50 queries should be executed.
 
Looks that some same template variables are reloaded 2-4 times, often only difference is 1 or second difference in start/end times.

This query is executed 3 or 4 times
`
label_values(datasource_seconds_count{job=~"$job", country=~"$country", dnsname=~"$appall"}, statementType) 
`
![image](https://user-images.githubusercontent.com/19704268/181798221-80b5a375-7703-4de3-b504-7e9ab91ffe01.png)

This query 2 times
`
label_values(dataSource_connections_active{job=~"$job", dnsname=~"$appall"}, dnsname)
`
![image](https://user-images.githubusercontent.com/19704268/181798960-b37742de-3942-436a-8e02-e2c9e8701f7d.png)

This query only once
`
label_values(job_le:http_server_requests_seconds_bucket:2xx_rate1m{job=~"$job"}, le)
`

There is probably some issues in template variable dependency management due common variables used in queries, either some variables are reloaded/flushed and thus result reload of dependent variables too, or multiple query generated when more than one dependent variable depending when variable is ready. Might be related to facts that queries might get stalled for few second.

Should I create separate bug ?

**Comment by natellium** [comment]
Yes, please @jtorkkel let's create a separate issue for it.

**Comment by younsl** [comment]
I hope this comment helps.

In my case, I had an issue where the Grafana dashboard would hang indefinitely when using an outdated Prometheus data source. Also, variables were not working correctly in the dashboard.

Even though the original metrics stored in Prometheus were properly collected and all referenced metric labels were correct, the issue still occurred. The root cause was a significant version gap between Grafana and Prometheus. [Prometheus 2.7.1](https://github.com/prometheus/prometheus/releases/tag/v2.7.1), released on January 31, 2019, was quite outdated, while [Prometheus 3.2.0](https://github.com/prometheus/prometheus/releases/tag/v3.2.0), released on February 17, 2025, included many updates and optimizations. This large gap likely contributed to the issue.

- Environment where the issue happened: Grafana 11.5.2 + Prometheus 2.7.1

After upgrading Prometheus from 2.7.1 to 3.2.0, the hanging issue in the Grafana dashboard was fixed. This shows that if the version gap between Prometheus and Grafana is too big, unexpected problems can happen. Keeping both updated helps prevent these issues.
